src/singlelayerqg.jl

```diff
@@ -0,0 +1,508 @@
+module SingleLayerQG
+
+export
+  Problem,
+  set_q!,
+  updatevars!,
+
+  energy,
+  kinetic_energy,
+  potential_energy,
+  energy_dissipation,
+  energy_work,
+  energy_drag,
+  enstrophy,
+  enstrophy_dissipation,
+  enstrophy_work,
+  enstrophy_drag
+
+using
+  CUDA,
+  Reexport
+
+@reexport using FourierFlows
+
+using FFTW: rfft
+using LinearAlgebra: mul!, ldiv!
+using FourierFlows: parsevalsum, parsevalsum2
+
+nothingfunction(args...) = nothing
+
+"""
+    Problem(; parameters...)
+
+Construct a SingleLayerQG problem.
+"""
+
+function Problem(dev::Device=CPU();
+  # Numerical parameters
+                  nx = 256,
+                  ny = nx,
+                  Lx = 2π,
+                  Ly = Lx,
+                  dt = 0.01,
+  # Physical parameters
+                   β = 0.0,
+  deformation_radius = Inf,
+                 eta = nothing,
+  # Drag and (hyper-)viscosity
+                   ν = 0.0,
+                  nν = 1,
+                   μ = 0.0,
+  # Timestepper and equation options
+             stepper = "RK4",
+               calcF = nothingfunction,
+          stochastic = false,
+                   T = Float64)
+
+  # the grid
+  grid = TwoDGrid(dev, nx, Lx, ny, Ly; T=T)
+  x, y = gridpoints(grid)
+
+  # topographic PV
+  eta === nothing && ( eta = zeros(dev, T, (nx, ny)) )
+
+  params = deformation_radius == Inf ? BarotropicQGParams(grid, T(β), eta, T(μ), T(ν), nν, calcF) : EquivalentBarotropicQGParams(grid, T(β), T(deformation_radius), eta, T(μ), T(ν), nν, calcF)
+
+  vars = calcF == nothingfunction ? Vars(dev, grid) : (stochastic ? StochasticForcedVars(dev, grid) : ForcedVars(dev, grid))
+
+  equation = Equation(params, grid)
+
+  return FourierFlows.Problem(equation, stepper, dt, grid, vars, params, dev)
+end
+
+
+# ----------
+# Parameters
+# ----------
+
+"""
+    Params(g::TwoDGrid, β, eta, μ, ν, nν, calcF)
+
+Returns the params for an unforced two-dimensional barotropic QG problem.
+"""
+abstract type SingleLayerQGParams <: AbstractParams end
+
+struct Params{T, Aphys, Atrans, ℓ} <: SingleLayerQGParams
+                   β :: T            # Planetary vorticity y-gradient
+  deformation_radius :: ℓ            # deformation radius
+                 eta :: Aphys        # Topographic PV
+                etah :: Atrans       # FFT of Topographic PV
+                   μ :: T            # Linear drag
+                   ν :: T            # Viscosity coefficient
+                  nν :: Int          # Hyperviscous order (nν=1 is plain old viscosity)
+              calcF! :: Function     # Function that calculates the forcing on QGPV q
+end
+
+const BarotropicQGParams = Params{<:AbstractFloat, <:AbstractArray, <:AbstractArray, Nothing}
+const EquivalentBarotropicQGParams = Params{<:AbstractFloat, <:AbstractArray, <:AbstractArray, <:AbstractFloat}
+
+get_topographicPV_grid_values(eta::Function, grid::AbstractGrid{T, A}) where {T, A} = A([eta(grid.x[i], grid.y[j]) for i=1:grid.nx, j=1:grid.ny])
```

Comment:
Maybe out of scope, but this works for any function of x, y (not just topographic PV), so you could put this in FourierFlows `utils.jl` or something and instead call it

```julia
on_grid(func, grid::TwoDimGrid{T, A}) where {T, A} = ...
```

Reply:
Good suggestion. Let's do it in a separate PR.